File: src/main/java/org/taktik/icure/dao/impl/PropertyDAOImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao.impl;

import org.ektorp.support.View;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Repository;
import org.taktik.icure.dao.PropertyDAO;
import org.taktik.icure.dao.impl.idgenerators.IDGenerator;
import org.taktik.icure.dao.impl.ektorp.CouchDbICureConnector;
import org.taktik.icure.entities.Property;

import java.util.List;

@Repository("propertyDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.Property' && !doc.deleted) emit(doc._id )}")
public class PropertyDAOImpl extends CachedDAOImpl<Property> implements PropertyDAO {
	private static final String IDENTIFIER = "type.identifier";

	@Autowired
    public PropertyDAOImpl(@SuppressWarnings("SpringJavaAutowiringInspection") @Qualifier("couchdbConfig") CouchDbICureConnector couchdb, IDGenerator idGenerator, @Qualifier("entitiesCacheManager") CacheManager cacheManager) {
        super(Property.class, couchdb, idGenerator, cacheManager);
        initStandardDesignDocument();
    }

    @Override
    @View(name = "by_identifier", map = "classpath:js/property/By_identifier_Map.js")
    public Property getByIdentifier(String propertyIdentifier) {
		Cache.ValueWrapper wrappedValue = getWrapperFromCache("PID:"+propertyIdentifier);
        if (wrappedValue == null) {
            List<Property> result = queryView("by_identifier", propertyIdentifier);
			Property value = result != null && result.size() == 1 ? result.get(0) : null;

            if (value!=null && value.getId()!=null) { putInCache(value.getId(), value); }
			return value;
        }
        return (Property)wrappedValue.get();
    }

	@Override
	public void evictFromCache(Property entity) {
        super.evictFromCache(entity);
        cache.evict("PID:"+entity.getType().getIdentifier());
    }

	@Override
	public void putInCache(String key, Property entity) {
		super.putInCache(key, entity);
		cache.put("PID:"+entity.getType().getIdentifier(), entity);
	}
}

File: src/main/kotlin/org/taktik/icure/asyncdao/impl/PropertyDAOImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao.impl

import kotlinx.coroutines.flow.map
import org.taktik.couchdb.annotation.View
import org.springframework.beans.factory.annotation.Qualifier

import org.springframework.stereotype.Repository
import org.taktik.couchdb.queryViewIncludeDocs
import org.taktik.icure.asyncdao.PropertyDAO
import org.taktik.couchdb.id.IDGenerator
import org.taktik.icure.entities.Property
import org.taktik.icure.properties.CouchDbProperties
import org.taktik.icure.spring.asynccache.AsyncCacheManager

import org.taktik.icure.utils.firstOrNull

@Repository("propertyDAO")
@View(name = "all", map = "function(doc) { if (doc.java_type == 'org.taktik.icure.entities.Property' && !doc.deleted) emit(null, doc._id )}")
class PropertyDAOImpl(couchDbProperties: CouchDbProperties,
                      @Qualifier("configCouchDbDispatcher") couchDbDispatcher: CouchDbDispatcher, idGenerator: IDGenerator, @Qualifier("asyncCacheManager") asyncCacheManager: AsyncCacheManager) : CachedDAOImpl<Property>(Property::class.java, couchDbProperties, couchDbDispatcher, idGenerator, asyncCacheManager), PropertyDAO {

    @View(name = "by_identifier", map = "classpath:js/property/By_identifier_Map.js")
    override suspend fun getByIdentifier(propertyIdentifier: String): Property? {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)

        val wrappedValue = getWrapperFromCache("PID:$propertyIdentifier")
        if (wrappedValue == null) {
            val result = client.queryViewIncludeDocs<String, String, Property>(createQuery(client, "by_identifier").includeDocs(true).key(propertyIdentifier)).map { it.doc }.firstOrNull()
            if (result?.id != null) {
                putInCache(result.id, result)
            }
            return result
        }
        return wrappedValue.get() as Property
    }

    override suspend fun evictFromCache(entity: Property) {
        super.evictFromCache(entity)
    }

    override suspend fun putInCache(key: String, entity: Property) {
        super.putInCache(key, entity)
    }

    companion object {
        private val IDENTIFIER = "type.identifier"
    }
}
